<img src="https://i.ibb.co/TcVMz44/logo.jpg" alt="logo" border="0" width=200>

# Computational Astrophysics 2021
---
## Eduard Larrañaga

Observatorio Astronómico Nacional\
Facultad de Ciencias\
Universidad Nacional de Colombia

---

### About this Notebook

In this notebook we use `python` and the `AstroPy` package to illustrate some basic concepts of astrophysics such as distances and brightness. 

---

## The `AstroPy` Project

The `AstroPy` project is a core package for astronomy and astrophysics using `Python`. It includes a lot of functions that we will use along this course.

`AstroPy` is included in GoogleColab and in Anaconda. To update to the latest version, use

`conda update astropy`

This package includes module with physical constants and a module to manipulate units. The complete list of constants included can be found at

https://docs.astropy.org/en/stable/constants/index.html

and the complete list of units included can be found at

https://docs.astropy.org/en/stable/units/index.html

In [ ]:
import numpy as np
import astropy.units as u
from astropy import constants as const

Newton's gravitational constant is

In [ ]:
const.G

<<class 'astropy.constants.codata2018.CODATA2018'> name='Gravitational constant' value=6.6743e-11 uncertainty=1.5e-15 unit='m3 / (kg s2)' reference='CODATA 2018'>

Speed of light in vacuum is

In [ ]:
const.c

<<class 'astropy.constants.codata2018.CODATA2018'> name='Speed of light in vacuum' value=299792458.0 uncertainty=0.0 unit='m / s' reference='CODATA 2018'>

Mass of the Sun

In [ ]:
const.M_sun

<<class 'astropy.constants.iau2015.IAU2015'> name='Solar mass' value=1.988409870698051e+30 uncertainty=4.468805426856864e+25 unit='kg' reference='IAU 2015 Resolution B 3 + CODATA 2018'>

Reduced Planck's constant, $\hbar$, and Boltzmann's constant, $k_B$, are

In [ ]:
const.hbar

<<class 'astropy.constants.codata2018.CODATA2018'> name='Reduced Planck constant' value=1.0545718176461565e-34 uncertainty=0.0 unit='J s' reference='CODATA 2018'>

In [ ]:
const.k_B

<<class 'astropy.constants.codata2018.CODATA2018'> name='Boltzmann constant' value=1.380649e-23 uncertainty=0.0 unit='J / K' reference='CODATA 2018'>

Information about a particular constant can be obtained as

In [ ]:
print(const.c)

  Name   = Speed of light in vacuum
  Value  = 299792458.0
  Uncertainty  = 0.0
  Unit  = m / s
  Reference = CODATA 2018


The value of a constant can be given in other units,

In [ ]:
const.c.to('pc/yr')

<Quantity 0.30660139 pc / yr>

In [ ]:
const.c.to('lyr/yr')

<Quantity 1. lyr / yr>

It is possible to define a mass equal of $1$ solar mass

In [ ]:
M = 1*u.Msun
M

<Quantity 1. solMass>

and we can calculate the corresponding Schwarzschild radius using the imported constants,

In [ ]:
G = const.G
c = const.c
rS = 2*G*M/c**2
rS

<Quantity 1.48523205e-27 m solMass / kg>

Note that the solar mass units can be transformed into kg to obtain a radius in units of meters. This is acomplished by the method `.decompose()`, 

In [ ]:
rS = rS.decompose()
rS

<Quantity 2953.2500761 m>

and it is transformed into other units with the method `.to()`

In [ ]:
rS.to(u.km)

<Quantity 2.95325008 km>



---
## Distances in Astrophysics 

**`Carroll and Ostlie (2017). Section 3.1`**

The first concept in astrophysics that we will present is that of distance. Here, we will present some units of measurement of distances used in astronomy and astrophysics through the package `astropy.units`. A complete list of units in the package can be found at
https://docs.astropy.org/en/stable/units/index.html

First, we introduce the **light-year** (lyr) as the distance travelled by light in vacuum in one Julian year (i.e. 365.25 days):

In [ ]:
ly = 1*u.lyr
ly

<Quantity 1. lyr>

This quantity can be written in other units such as meters or kilometers by usint the `.to()` method,

In [ ]:
ly.to(u.m)

<Quantity 9.46073047e+15 m>

Due to our location in the Solar System , it is usual that all distances are given with respect to the Sun and that the unit of measurement at this scale is chosen as the mean distance Sun-Earth. This is called an **Astronomical Unit** (au),

In [ ]:
au = 1*u.au
au

<Quantity 1. AU>

$1$ au is equivalent to

In [ ]:
au.to(u.m)

<Quantity 1.49597871e+11 m>

In [ ]:
au.to(u.lyr)

<Quantity 1.58125074e-05 lyr>

### Parallax and the definition of parsec ###

One of the initial forms of measurment of distances to stars in the Milky Way si known as the **parallax method**, in which it is used the annual shift of the angular position of an object in the sky. Using basic trigonometry, the distance to a satr is given by

\begin{equation}
d = \frac{b}{\tan \alpha}
\end{equation}

where $b$ is the baseline of the triangle and $\alpha$ is the parallax angle. For measurements from Earth, the baseline is chosen as $1$ au.

Now, we define a function to calculate the distance for a given parallax angle,

In [ ]:
def d(alpha=1*u.arcsec, baseline = 1*u.au):
    '''
    ------------------------------------------
    d(alpha, baseline)
    ------------------------------------------
    Returns the distance to a star using the 
    parallax method.
    
    Arguments:
    baseline in au
    parallax angle alpha given in arsec.
    ------------------------------------------
    '''
    return baseline/np.tan(alpha)

#### Example. $\alpha$-Centauri ####
The closest star system to the Sun is called $\alpha$-Centauri. It is a triple system, consisting of three stars: $\alpha$-Centauri A (Rigil Kentaurus), $\alpha$-Centauri B (Toliman) and $\alpha$-Centauri C (Proxima Centauri). The parallax angle for this system is 754.81 milli-arsec. Therefore, it is located at a distance of 

In [ ]:
alphaCentauri_distance = d(0.75481*u.arcsec)
alphaCentauri_distance

<Quantity 273267.18809525 AU>

This is clearly a large value but it can be converted to light-years,

In [ ]:
alphaCentauri_distance.to(u.lyr)

<Quantity 4.32103944 lyr>

---
The parallax method permits to define a distance unit called **parsec** (pc), corresponding to the distance at which $1$ au subtends a parallax angle of $1$ arcsec. This unit is equivalent to

In [ ]:
pc = d(1*u.arcsec)
pc

<Quantity 206264.80624548 AU>

In [ ]:
pc.to(u.pc)

<Quantity 1. pc>

In [ ]:
pc.to(u.lyr)

<Quantity 3.26156378 lyr>

Today, parallax angles can be measured up to the order of milli-arcsec. This corresponds to distance of the order of kpc,

In [ ]:
d(1*u.marcsec).to(u.pc)

<Quantity 1000. pc>

For far objects, for example at distances of the order of Mpc, the parallax method can not be used. Instead, methods such as relative brightness are used to estimate distances.

#### Example. $\alpha$-Centauri ####
The distance to $\alpha$-Centauri in pc is

In [ ]:
alphaCentauri_distance.to(u.pc)

<Quantity 1.32483671 pc>

#### Example. 61 Cygni ####
After 4 years of observation, in 1838 Bessel reported a parallax angle of $\alpha=0.316$ arcsec for the star 61-Cygni. The distance to this star is 

In [ ]:
cygni_distance = d(0.316*u.arcsec)
cygni_distance 

<Quantity 652736.72862954 AU>

In [ ]:
cygni_distance.to(u.pc)

<Quantity 3.16455696 pc>

---
## The Continuous Spectrum of Light
**`Carroll and Ostlie (2017). Section 3.2`**

### Apparent Magnitude ###

The magnitude scale defined by the Hipparchus used numbers from $1$ to $6$ where $1$ corresponds to the brightest objects. Hence, the magnitude scale difference in the visual range is 5. This is known as the apparent magnitude ($m$).
Today, we know that the magnitude scale difference $\Delta m = 5$ corresponds to a brightness (luminous energy) ratio of about 100. This means that magnitudes are logaritmic measures of brightness (similar to the case of audio loudness in which decibels define a base 10 logarithmic scale). 

Concerning magnitudes, we can obtain the logarithm base $b$ of the scale by solving the equation

\begin{equation}
b^{\Delta m} = b^5 = 100
\end{equation}

The `sympy` functions `Symbol` and `solve` gives the solution,

In [ ]:
from sympy import Symbol, solve
b = Symbol('b')
solve(100.-b**5, b)

[2.51188643150958,
 -2.03215881103109 - 1.4764497998749*I,
 -2.03215881103109 + 1.4764497998749*I,
 0.776215595276303 - 2.38894595888057*I,
 0.776215595276303 + 2.38894595888057*I]

The real value corresponds to the logarithm base $b=2.51188643150958$. 

### Flux and Luminosity

The modern measurement of "brightness" is given in terms of the **radiant flux** $F$, defined as the total amount of energy (in all wavelengths) crossing a unit area oriented perpendicular to the direction of the light ray per unit time. It is measured in Watts (Joules per second).

$F$ depends on the **intrinsec luminosity** of the star, $L$, and the distance from the observer. The inverse square law for light states that

\begin{equation}
F = \frac{L}{4\pi r^2}
\end{equation}

The relation between flux and apparent magnitude is given by considering that a difference of $\Delta m = m_1 - m_2 = 5$ correspond to a ratio of $100$ in the fluxes, i.e.

\begin{equation}
\frac{F_2}{F_1} = 100^{\frac{\Delta m}{5}}.
\end{equation}

or inverting

\begin{equation}
\Delta m = m_1 - m_2 = - 2.5 \log_{10} \left( \frac{F_1}{F_2} \right)
\end{equation}


Some examples of apparent magnitudes of stars are:

| *Star*  |  $m$  |
|:-------:|:-----:|
| Sun     | -27   |
|Sirius   | -1.46 |
|Arcturus | -0.04 |
|Aldebaran| 0.85  |
|Spica    | 1.04  |
|Procyon  | 0.34  |


### Absolute Magnitude ###

The absolute magnitude ($M$) describes the intrinsic luminosity emitted by an object and is defined to be equal to the apparent magnitude that the object would have if it were placed at a certain distance from Earth. Since there are no major stars within 1 parsec from the Sun, but there are many important stellar objects within 10 parsecs, it has been historically chosen a reference distance of 10 pc for stars. There exist a more complex definition of absolute magnitude for planets and otrher Solar System bodies.

Then, the absolute magnitude is obtained from the apparent magnitude through

\begin{equation}
M = m - 5(\log_{10} D -1)
\end{equation}

where $D$ stands for the object's distance measured in parsecs. Solving for $D$, we can write

\begin{equation}
D = 10^\frac{m - M + 5}{5}
\end{equation}

#### Example. Absolute Magnitude of the Sun 

The Sun is located at a distance of $1$ au from Earth and its apparent magnitude is $m=-26.83$. Lets calculate its absolute magnitude

In [ ]:
def M(m,D):
    '''
    --------------------------------------------
    M(m,D)
    --------------------------------------------
    Returns the absolute magnitude of a star.
    
    Arguments:
    m: apparent magnitude
    D: Distance to the Sun in parsecs
    --------------------------------------------
    '''
    return m - 5*(np.log10(D) - 1)


In [ ]:
D_sun = 1*u.au.to(u.pc)

M(-26.83, D_sun)

4.742125665882298

---
## Color Index
**`Carroll and Ostlie (2017). Section 3.6`**

Since actual technology permits to register wavelengths outside the range of human vision, magnitudes are estimated using various filters.

The standarized UBV system introduced by Johnson and Morgan uses three filters: ultraviolet, blue and visible and therefore one can measure three apparent magnitude ranges: $m_U$, $m_B$ and $m_V$. These filters have the ranges:

- U (ultraviolet): centered at $365$ nm with an effective bandwidth of $68$ nm

- B (blue): centered at $440$ nm with an effective bandwidth of $98$ nm

- V (visual): centered at $550$ nm with an effective bandwidth of $89$ nm


When the distance from the star $D$ is known, it is possible to obtain the absolute magnitudes $M_U$, $M_B$ and $M_V$.

The **color index** is defined as the difference between a pair of these magnitudes. For example, the **U-B color index** is defined as

\begin{equation}
U-B= M_U - M_B,
\end{equation}

while the **B-V color index** is defined as

\begin{equation}
B-V= M_B - M_V.
\end{equation}

*Since magnitudes decrease with increasing brightness, a star with smaller B-V color index is bluer than one with a larger B-V color index.*

#### Example. Color index for Sirius 

Sirius is the brightest star in the night sky. Its U, B and V apparent magnitudes are:

In [ ]:
U = -1.47
B = -1.43
V = -1.44

Then, the U-B color index and the B-V color index for Sirius are

In [ ]:
U-B 

-0.040000000000000036

In [ ]:
B-V

0.010000000000000009